src/IW/Sync/Search.hs

```diff
@@ -28,70 +35,117 @@ import qualified Data.Text as T
 import qualified Data.Vector as V
 
 
--- | Fetch all repositories built with the Haskell language.
-fetchAllHaskellRepos
-    :: ( MonadIO m
+githubSearch
+    :: forall a env m.
+       ( MonadIO m
        , WithError m
        , WithLog env m
+       , FromJSON a
+       , Typeable a
        )
-    => m [GitHub.Repo]
-fetchAllHaskellRepos = liftGitHubSearchToApp searchHaskellRepos
+    => [Text] -- ^ Paths
+    -> Text   -- ^ Query String
+    -> Day    -- ^ From day
+    -> Day    -- ^ To day
+    -> Int    -- ^ Page
+    -> m [a]
+githubSearch paths queryString from to page = liftIO (executeRequest' $ query paths queries) >>= \case
+    Left err -> throwError $ githubErrToAppErr err
+    Right (SearchResult count vec) -> do
+        log Info $ "Query executed with the following paths and query parameters: " <> show paths <> " " <> show queries
+        log Info $ "Fetched total of " <> show count <> " " <> typeName @a <> "s..."
+        searchLimit <- getSearchRateLimit
+        log Info $ "Search rate limit information: " <> show searchLimit
+        pure $ V.toList vec
+  where
+    queries :: [(ByteString, Maybe ByteString)]
+    queries = pagination <> [("q", Just $ encodeUtf8 $ strip queryString <> " " <> dateRange from to)]
 
--- | Convert a value of the @GitHub.Repo@ type to a value of our own @Repo@ type.
-fromGitHubRepo :: GitHub.Repo -> Repo
-fromGitHubRepo githubRepo = Repo
-    { repoOwner      = RepoOwner $ GitHub.untagName $ GitHub.simpleOwnerLogin $ GitHub.repoOwner githubRepo
-    , repoName       = RepoName  $ GitHub.untagName $ GitHub.repoName githubRepo
-    , repoDescr      = fromMaybe "" $ GitHub.repoDescription githubRepo
-    , repoCategories = SqlArray []
-    }
+    pagination :: [(ByteString, Maybe ByteString)]
+    pagination =
+        [ ("per_page", Just "100")
+        , ("page", Just $ show page)
+        ]
+
+    dateRange :: Day -> Day -> Text
+    dateRange from' to' = "created:" <> julianDayToIso from' <> ".." <> julianDayToIso to'
+
+    julianDayToIso :: Day -> Text
+    julianDayToIso = fromString . formatTime defaultTimeLocale (iso8601DateFormat Nothing)
+
+getSearchRateLimit
+    :: forall m.
+       ( MonadIO m
+       , WithError m
+       )
+    => m Limits
```

Comment:
Signature for this function can be written in one line. No need to put every argument on a separate line if function fits into single line. Having short functions with short type signatures helps readability better rather than an unnecessary split of type signature into multiple lines.

Reply:
@chshersh Yes for sure, I went too far with this one :sweat_smile: